Add a mnist example that runs a flax model on TFLite using jax2tf. (#5121)
google/jax: jax/experimental/jax2tf/examples/tflite/mnist/README.md

> # MNIST classification

Reviewer: (Never mind - I see you OKed the previous change 👍 )

> This directory contains the code necessary to train a simple convolution network on the MNIST dataset, and subsequently convert it using jax2tf into a model that can be converted to tflite format.

Reviewer: Suggestion (just making the complex sentence more accessible for readability):

```suggestion
This directory contains the code that demonstrates how to:

1. Train a simple convolutional neural network on the MNIST dataset; and
2. Use jax2tf — that helps stage JAX programs out as TensorFlow graphs — into a TensorFlow [SavedModel](https://www.tensorflow.org/guide/saved_model) that can be used again for conversion into a [TensorFlow Lite (TF Lite)](https://www.tensorflow.org/lite/) format for on-device machine learning inference.
```

@qiuminxu Let me know if I got this gist of this right 😃 (especially the SavedModel part)

> This example is based on the [TensorFlow Lite demo of building a handwritten digit classifier app with TensorFlow Lite](https://developer.android.com/codelabs/digit-classifier-tflite). The model code is taken from [FLAX Linen Mnist Example](https://github.com/google/flax/tree/master/linen_examples/mnist).

Reviewer: Suggestion (emphasis on Android, etc):

```suggestion
This example is based on the [TensorFlow Lite demo](https://developer.android.com/codelabs/digit-classifier-tflite) of building a handwritten digit classifier app for Android with TensorFlow Lite. The model training code is based on the [Flax Linen example](https://github.com/google/flax/tree/master/linen_examples/mnist) for classification on the MNIST dataset.
```

LMKWYT @qiuminxu

## Requirements

> * TensorFlow dataset mnist will be downloaded and prepared automatically, if necessary

Reviewer: Suggestion: "preprocessed" (instead of "prepared")

```suggestion
* This example uses [Flax](http://github.com/google/flax) and [TensorFlow Datasets](https://www.tensorflow.org/datasets). The code downloads the MNIST dataset from TensorFlow Datasets and preprocesses it before feeding the data into the neural network as input.
```

LMKWYT @qiuminxu (I'm actually working on a Flax MNIST walkthrough right now 😺)

## Training the model

> You can train and export the model to tflite format yourself running
> ```bash
> $ python mnist.py
> ```

Reviewer: Suggestion (I may be wrong about the code style you use here though but do add a blank line before the code block and reformat some words):

@qiuminxu LMKWYT of this (hope I got the gist right - essentially, `mnist.py` performs the entire thing end-to-end including quantization, right? I hope I got it right 😄 Let me know if I'm incorrect)

```suggestion
To train the model in JAX with the Flax library, convert it into a TensorFlow SavedModel, and export it into the TF Lite format, run the following command:

```shell
python mnist.py
```
```

Reviewer: @qiuminxu 

I think the rendering of triple backticks in my previous suggestion here caused a set of them to "fall off" - and I didn't spot it:

<img width="200" alt="image" src="https://user-images.githubusercontent.com/19637339/101399698-bc13e580-38c7-11eb-815c-0bbffe2cb4d5.png">

As a result, there's a small rendering issue on the changed Markdown:

<img width="500" alt="image" src="https://user-images.githubusercontent.com/19637339/101399685-b61e0480-38c7-11eb-9e5c-6618f321bf48.png">

Oops!

Do you mind adding the backticks back to close the code block?

    ```shell
    python mnist.py
    ```

🙏 Thanks!!

> The training itself takes only about 1 minute, assuming the dataset has already been downloaded. The dataset will be downloaded directly into a data/ directory in the /tmp/jax2tf/mnist directory. The training loop runs for 3 epochs, and the model is then converted to TensorFLow Lite and subsequently saved into /tmp/jax2tf/mnist/mnist.tflite.

Reviewer: Suggestion:
- Add "for demonstration purposes" since 3 loops are probably not enough, but I don't know 🤷‍♂️ 

```suggestion
The training itself should take about 1 minute, assuming the dataset has already been downloaded. The dataset is loaded directly into a `data/` directory under `/tmp/jax2tf/mnist`. 

This example's training loop runs for 3 epochs for demonstration purposes. After training, the model is converted to SavedModels and then — TF Lite. In the end, it is saved as `mnist.tflite` under `/tmp/jax2tf/mnist/`.
```

WDYT? @qiuminxu

> ### Convert to TF function using jax2tf converter
> Write a predict function from the trained model and use the jax2tf converter to convert it to a TF function.

Reviewer: ```suggestion
### Convert to TensorFlow function using the jax2tf converter

You can write a prediction function from the trained model and use the jax2tf converter to convert it to a TF function as follows:
```

> ```

Reviewer: Suggestion: for improved rendering

```suggestion
```python
```

def predict(image):
   return CNN().apply({'params': optimizer.target}, image)
```

> ```

Reviewer: Suggestion: for improved rendering

```suggestion
```python
```

> # Convert Flax model to TF function.

Reviewer: Small suggestion:

```suggestion
# Convert your Flax model to TF function.
```

(or use "the")

tf_predict = tf.function(
    jax2tf.convert(predict, enable_xla=False),
    input_signature=[
        tf.TensorSpec(shape=[1, 28, 28, 1], dtype=tf.float32, name='input')
    ],
    autograph=False)
```

> ### Convert to TensorFlow Lite
> [TensorFlow Lite converter](https://www.tensorflow.org/lite/convert#python_api_) provides a number of ways to convert a TensorFlow model to the TensorFlow Lite format. Here we use the `from_concrete_functions` API.

Reviewer: Suggestion (having pre-defined TensorFlow Lite as TF Lite before):
- Also, using the second person ("you") (a suggestion from the Google style guide)

```suggestion
### Convert the trained model to the TF Lite format

The [TF Lite converter](https://www.tensorflow.org/lite/convert#python_api_) provides a number of ways to convert a TensorFlow model to the TF format. For example, you can  use the `from_concrete_functions` API as follows:
```

> ```

Reviewer: ```suggestion
```python
```

> # Convert TF function to TF Lite format.

Reviewer: ```suggestion
# Convert your TF function to the TF Lite format.
```

converter = tf.lite.TFLiteConverter.from_concrete_functions(
    [tf_predict.get_concrete_function()])

converter.target_spec.supported_ops = [
    tf.lite.OpsSet.TFLITE_BUILTINS,  # enable TensorFlow Lite ops.
    tf.lite.OpsSet.SELECT_TF_OPS  # enable TensorFlow ops.
]
tflite_float_model = converter.convert()
```

> Note that tf.lite.OpsSet.SELECT_TF_OPS are required here to run regular TensorFlow ops (as compared to TensorFlow Lite ops) in the TensorFlow Lite runtime.

Reviewer: Suggestion: use "code" font with backticks

```suggestion
Note that the use of `tf.lite.OpsSet.SELECT_TF_OPS` in supported ops is required here to run regular TensorFlow ops (as compared to TensorFlow Lite ops) in the TF Lite runtime.
```

### Apply quantization

> The converted format is no different than one converted from a regular TensorFlow Model. We can apply quantization as usual.

Reviewer: Suggestions:
- I'd not use "as usual" for those who are new to on-device stuff - instead, I suggest to include a link to quantization if someone wants to study it
- Because the steps are sequential, I suggest to include "and now the next step/stage is...":

```suggestion
The next step is to perform [quantization](https://www.tensorflow.org/lite/performance/post_training_quantization). Because the converted format is no different than the one converted from a regular TensorFlow SavedModel, you can apply quantization:
```

LMKWYT @qiuminxu

> ```

Reviewer: ```suggestion
```python
```

# Re-convert the model to TF Lite using quantization.
converter.optimizations = [tf.lite.Optimize.DEFAULT]
tflite_quantized_model = converter.convert()
```

> ## Deploy the TensorFlow Lite model to Android App

Reviewer: Suggestion (having predefined TensorFlow Lite as TF Lite):

```suggestion
## Deploy the TF Lite model to your Android app
```

> Now you can follow the steps in the [Digit Classifier TensorFlow Lite example](https://developer.android.com/codelabs/digit-classifier-tflite) to deploy the App.

Reviewer: Suggestion (using sentence case here as per the Google style guide):

```suggestion
To deploy the TF Lite model to your Android app, follow the instructions in the [Build a handwritten digit classifier app](https://developer.android.com/codelabs/digit-classifier-tflite) codelab.
```

> When using a TensorFlow Lite model that has been converted with support for select TensorFlow ops, the client must also use a TensorFlow Lite runtime that includes the necessary library of TensorFlow ops. For this demo, you can use the standard TensorFlow AAR file, by specifying the [nightly dependencies](https://www.tensorflow.org/lite/guide/ops_select#android_aar).

Reviewer: Suggestion:
- Add the AAR definition in a link

```suggestion
When using a TF Lite model that has been converted with the support for select TensorFlow ops, the client must also use the TF Lite runtime that includes the necessary library of TensorFlow ops. In this example, you can use the standard TensorFlow [AAR](https://developer.android.com/studio/projects/android-library#aar-contents) file, by specifying the [nightly dependencies](https://www.tensorflow.org/lite/guide/ops_select#android_aar).
```

```
dependencies {
    implementation 'org.tensorflow:tensorflow-lite:0.0.0-nightly'
    // This dependency adds the necessary TF op support.
    implementation 'org.tensorflow:tensorflow-lite-select-tf-ops:0.0.0-nightly'
}
```

> You can also specify abiFilters to reduce the size of the TF Op dependencies.

Reviewer: Suggestion - code font:

```suggestion
You can also specify `abiFilters` to reduce the size of the TensorFlow op dependencies.
```

```
android {
    defaultConfig {
        ndk {
            abiFilters 'armeabi-v7a', 'arm64-v8a'
        }
    }
}
```

## Limitations

> * TensorFlow Lite runtime doesn't support XLA ops, therefore you need to specify `enable_xla=False` to use the jax2tf conversion purely based on TensorFlow (not XLA). The support is limited to a subset of jax ops. 

Reviewer: Suggestion:

```suggestion
* The TF Lite runtime currently doesn't support XLA ops. Therefore, you need to specify `enable_xla=False` to use the jax2tf conversion purely based on TensorFlow (not XLA). The support is limited to a subset of JAX ops.
```
